packages/discord.js/src/client/websocket/WebSocketManager.js

```diff
@@ -152,156 +162,108 @@ class WebSocketManager extends EventEmitter {
-        this.shardQueue.add(shard);
-      } else {
-        throw error;
+          this.checkShardsReady();
+        });
+        shard.status = Status.Connecting;
       }
     }
-    // If we have more shards, add a 5s delay
-    if (this.shardQueue.size) {
-      this.debug(`Shard Queue Size: ${this.shardQueue.size}; continuing in 5 seconds...`);
-      await sleep(5_000);
-      return this.createShards();
-    }
 
-    return true;
+    await this._ws.connect();
   }
 
   /**
-   * Handles reconnects for this manager.
+   * Attaches event handlers to the internal WebSocketShardManager from `@discordjs/ws`.
    * @private
-   * @returns {Promise<boolean>}
    */
-  async reconnect() {
-    if (this.reconnecting || this.status !== Status.Ready) return false;
-    this.reconnecting = true;
-    try {
-      await this.createShards();
-    } catch (error) {
-      this.debug(`Couldn't reconnect or fetch information about the gateway. ${error}`);
-      if (error.httpStatus !== 401) {
-        this.debug(`Possible network error occurred. Retrying in 5s...`);
-        await sleep(5_000);
-        this.reconnecting = false;
-        return this.reconnect();
+  attachEvents() {
+    this._ws.on(WSWebSocketShardEvents.Debug, ({ message, shardId }) => this.debug(message, shardId));
+    this._ws.on(WSWebSocketShardEvents.Dispatch, ({ data, shardId }) => {
+      this.client.emit(Events.Raw, data, shardId);
+      const shard = this.shards.get(shardId);
+      this.handlePacket(data, shard);
+      if (shard.status === Status.WaitingForGuilds && WaitingForGuildEvents.includes(data.t)) {
+        shard.gotGuild(data.d.id);
       }
-      // If we get an error at this point, it means we cannot reconnect anymore
-      if (this.client.listenerCount(Events.Invalidated)) {
+    });
+
+    this._ws.on(WSWebSocketShardEvents.Ready, ({ data, shardId }) => {
+      this.shards.get(shardId).onReadyPacket(data);
+    });
+
+    this._ws.on(WSWebSocketShardEvents.Closed, ({ code, reason = '', shardId }) => {
+      if (UNRESUMABLE_CLOSE_CODES.includes(code) && this.destroyed) {
+        this.shards.get(shardId).status = Status.Disconnected;
         /**
-         * Emitted when the client's session becomes invalidated.
-         * You are expected to handle closing the process gracefully and preventing a boot loop
-         * if you are listening to this event.
-         * @event Client#invalidated
+         * Emitted when a shard's WebSocket disconnects and will no longer reconnect.
+         * @event Client#shardDisconnect
+         * @param {CloseEvent} event The WebSocket close event
+         * @param {number} id The shard id that disconnected
          */
-        this.client.emit(Events.Invalidated);
-        // Destroy just the shards. This means you have to handle the cleanup yourself
-        this.destroy();
+        this.client.emit(Events.ShardDisconnect, { code, reason, wasClean: true }, shardId);
```

Comment:
According to the types, `CloseEvent` (the first event parameter) should also have a `target`

https://github.com/discordjs/discord.js/blob/79f664e1d93c38d4c80e82cfa0fa774e2c41f045/packages/discord.js/typings/index.d.ts#L4859-L4864

However, according to the JSDoc, which is just a link to MDN, it does not have a `target`
https://developer.mozilla.org/en-US/docs/Web/API/CloseEvent#instance_properties

One of those should probably be changed